### ISSUE TITLE
Add support for Zip64 extraction

### ISSUE THREAD

**Issue by brendand** [abbeycode/UnzipKit#25]
I have a 6 GB zip file I need to uncompress and it gets about 75% the way through and then fails with an error:

UnzipKit error...
Name: Error reading a file in the archive
Detail: Error opening password protected archive (-103)
UnzipKit error...
Name: Parameter error
Detail: Error closing current file 'Photos/B52AEB4D-1E5B-4017-A311-E009658A5D9D.jpg'
Error uncompressing: Error Domain=UZKErrorDomain Code=-102 "Parameter error" UserInfo={NSUnderlyingError=0x608000c572b0 {Error Domain=UZKErrorDomain Code=107 "Error reading a file in the archive" UserInfo={NSLocalizedDescription=Error reading a file in the archive, NSLocalizedRecoverySuggestion=Error opening password protected archive (-103), NSURL=file:///Users/brendan/Downloads/TF-Susan%E2%80%99s%20MacBook%20Pro-2016-01-07T09-58-39.zip, NSLocalizedFailureReason=Error reading a file in the archive}}, NSLocalizedDescription=Parameter error, NSLocalizedRecoverySuggestion=Error closing current file 'Photos/B52AEB4D-1E5B-4017-A311-E009658A5D9D.jpg', NSURL=file:///Users/brendan/Downloads/TF-Susan%E2%80%99s%20MacBook%20Pro-2016-01-07T09-58-39.zip, NSLocalizedFailureReason=Parameter error}

I have no troubles uncompressing the file using the Finder or with the BetterZip app. But I can't get it to uncompress this 6 GB file.

The error says it's a password protected archive, but it's not actually.

If I delete that file, the next file in the list of files causes the error. So it seems like it might be a file size issue that UnzipKit maybe can't deal with files larger than 4 GB?

**Comment by abbeycode** [comment]
I haven't tried using UnzipKit with a file that big, but I'd love to see what we can do to get it to work for you.

To start with, can you give some sample code?

**Comment by brendand** [comment]
Thanks for your response!

This is the code I'm using. It's pretty straight forward:

```
NSFileManager *fileManager = [[NSFileManager alloc] init];
			
NSString *cachesDirectory = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) lastObject];
			
NSString *migrateFolder = [cachesDirectory stringByAppendingPathComponent:@"tapforms-migrate"];
			
NSError *error = nil;
			
success = [fileManager createDirectoryAtPath:migrateFolder
				 withIntermediateDirectories:YES
								  attributes:nil
									   error:&error];
			
if (!success) {
	TFFLog(@"Error creating folder at path: %@", migrateFolder);
}
			
UZKArchive *oldArchive = [[UZKArchive alloc] initWithURL:self.selectedBackupFileURL
												   error:&error];
progress.totalUnitCount = 100;
			
success = [oldArchive extractFilesTo:migrateFolder
						   overwrite:YES
							progress:^(UZKFileInfo * _Nonnull currentFile, CGFloat percentArchiveDecompressed) {
								dispatch_async(dispatch_get_main_queue(), ^{
									float percent = percentArchiveDecompressed * 100;
									progress.completedUnitCount = percent;
									progress.localizedAdditionalDescription = [currentFile.filename lastPathComponent];
								});
							} error:&error];
			
```

**Comment by brendand** [comment]
Also, the file I'm trying to unzip has about 945 items in it. Mostly JPG and TIF files, but some other file types and folders too.

**Comment by brendand** [comment]
Is there anything other information I can provide to you in order to help figure it out?

**Comment by abbeycode** [comment]
Thanks for what you provided. I'd love to be able to step through it myself to see what's going on. Could you post the archive someplace where I can download it? If you don't want to post the link here, you can DM it to me on Twitter (I'm @DovFrankel)

**Comment by brendand** [comment]
I sent you a DM. Hopefully you got it. From my @tapforms account.

**Comment by brendand** [comment]
Awesome. Thanks for looking into it!

**Comment by brendand** [comment]
Was just testing the big file with your latest commits and received this error that caused a crash:

```
2016-01-24 14:55:53.446 Tap Forms[24903:3612880] file selected: file:///Users/brendan/Downloads/TF-Susan%E2%80%99s%20MacBook%20Pro-2016-01-07T09-58-39.zip
2016-01-24 14:57:19.112 Tap Forms[24903:3614996] UnzipKit error...
Name: Error reading a file in the archive
Detail: Error opening archive (-103)
2016-01-24 14:57:19.112 Tap Forms[24903:3614996] UnzipKit error...
Name: Parameter error
Detail: Error closing current file (-102) 'Photos/B52AEB4D-1E5B-4017-A311-E009658A5D9D.jpg'
2016-01-24 14:57:19.112 Tap Forms[24903:3614996] *** -[NSError retain]: message sent to deallocated instance 0x600006857ac0
```

**Comment by abbeycode** [comment]
I'm aware of what's causing the crash (my continuous integrations started failing). I'll let you know when I have something for you to look at, to address your issue. So far I've improved the memory footprint and some error messages.

**Comment by brendand** [comment]
Awesome. Thanks for your hard work!

**Comment by brendand** [comment]
This last commit fixed the crash. Thanks!

So do you have an idea what is causing that big file not to be unzipped? It always stops on the same file. Right now it just shows this error:

Error Domain=UZKErrorDomain Code=107 "Error reading a file in the archive"

Is it because it has reached the 4 GB limit of the 32 bit memory space? Or is it some other reason?

Even if it was from a bad file, would it be possible to provide the error info so I can keep track of a list of bad files and then display that to the user, but still continue unzipping the rest of the archive?

Although that still doesn't explain why I can unzip it with the Finder and BetterZip.

**Comment by abbeycode** [comment]
It does look like it's a bad file (from what I can tell), or more specifically, a bad file header. I made an update to make sure the "Bad zip file" error shows up as an underlying error.

Regarding why Finder and BetterZip work, it could be that they're less strict. What's coming through as incorrect with that file is the "magic number", which should always be `0x04034b50`. Do you know how the file was created? Was it created with UnzipKit? If not, perhaps the app that created the file hit into some sort of size limit.

I'd be curious to see if the same issue manifests itself with that single file removed. When I removed it with BetterZip, it succeeded, but it re-created the whole file from scratch.

**Comment by brendand** [comment]
I'll try removing that file, but I'm pretty sure when I did that, it moved the problem to the next file. I'll try it again. The file was initially created using ZipArchive. I didn't know about UnzipKit before last week. ZipArchive isn't able to uncompress it either.

Is there a way to skip the file and continue unzipping?

**Comment by abbeycode** [comment]
Unfortunately, no, unless you were to extract each file yourself using another method. You can try `-[UZKArchive performOnDataInArchive:error:]` (which might also stop on an error) or calling `-[UZKArchive listFileInfo]` and then call `-[UZKArchive extractData:progress:error:]` on each one. That way would definitely work, since you could ignore any errors, but it's a little less abstracted than the `performOnData...` call.

**Comment by brendand** [comment]
Ah ok. At least there's a way then. Thanks!

**Comment by brendand** [comment]
Did you only find that one file with the problem? I deleted that file and re-saved in BetterZip and tried unzipping it again. But then got an error on the next file. I did that about 5 times, each time the next file down the list gets the error.

**Comment by abbeycode** [comment]
Okay, interesting. Can you give me screenshots of the options you're using in BetterZip? Once I save it with BetterZip, UnzipKit fails on the first file.

**Comment by brendand** [comment]
I'm just doing a regular save. Not using any of the built-in presets. I am using version 3.0.4 of BetterZip. 

<img width="687" alt="screen shot 2016-01-26 at 12 04 59 pm" src="https://cloud.githubusercontent.com/assets/586075/12591462/145704a6-c425-11e5-90ae-2ce93e6c83d3.png">

**Comment by abbeycode** [comment]
Ah, that explains it (probably). I'm still on BetterZip 2, and the new version adds "Direct Mode", which is what I'd like to use on this archive. While I work on upgrading it, can you try removing the file _before_ the one that throws the error? That should be `Photos/B0A180FB-7264-4895-93EC-720021DB01FC.tif`.

It basically seems like some file header, somewhere in the file, is off. Where it's looking for `Photos/B166E5DC-13BF-4DC4-AD71-0917B96FA893.tif` is not the right place in the file. I'm wondering if there may be something wrong with the earlier file, which is leaving subsequent ones at a bad offset.

**Comment by brendand** [comment]
That didn't work. I deleted even a few files before and after that file, then reserved, then tried unzipping again. I tried using direct mode and non-direct mode in BetterZip to delete those files.

**Comment by brendand** [comment]
Also, when I ran the BetterZip Test command to test the validity of the archive, it passed the test.

**Comment by abbeycode** [comment]
Okay, I've been digging through the hex of that file, and I've got _something_. I'm not sure what to do about it yet, but the central file header for the first file that fails has a hex offset for the file of `FF FF FF FF`, which is clearly the max value for the field. When it looks there, it's the middle of some file, not the beginning of the one it's expecting, which is why it fails. It looks like perhaps UnzipKit doesn't support Zip64 and this file requires it.

I'm still investigating.

**Comment by brendand** [comment]
Awesome. I'm glad you've found that. Ya, it seemed like it might have been a 64 bit issue to me since it always seemed to stop around the same spot in the file.

**Comment by brendand** [comment]
Bingo! You've got it! I was able to successfully decompress that gigantic file now.

One small glitch though, if you try to archive, an error pops up

```
/Users/brendan/Projects/iPhone/UnzipKit/Source/UZKArchive.m:404:58: error: implicit conversion increases floating-point precision: 'float' to 'CGFloat' (aka 'double') [-Werror,-Wdouble-promotion]
                        progress(info, bytesDecompressed / totalSize.floatValue);
                        ~~~~~~~~       ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
```

Changing that line to totalSize.doubleValue fixes the problem.

I can truly say that you have the best Zip/UnZip framework I have seen yet!

Thanks!

**Comment by abbeycode** [comment]
@brendand Thanks a lot. I'd like to address that compiler (error? warning?) but I'm not seeing it. How can I reproduce it?

**Comment by brendand** [comment]
It could be a new one in Xcode 7.3 (-Werror,-Wdouble-promotion) which I'm using now.

**Comment by abbeycode** [comment]
Ah, okay, I'm still on 7.2. 7.3 is still in beta, right?

**Comment by abbeycode** [comment]
Darn it, El Capitan is required for Xcode 7.3. Would you mind reporting that compiler warning as a separate issue? I'll address it once I've upgraded. It's not causing any issue for you at runtime, is it?

**Comment by brendand** [comment]
No. I simply changed it to doubleValue instead of floatValue and it's working perfectly. Sure, I'll create a separate issue.

**Comment by abbeycode** [comment]
This has been merged to `master`, and released to CocoaPods as v1.6.2. Thanks for pointing it out, and helping to solve it!
